### DISCUSSION TITLE
Code Naming and Refactoring

### DISCUSSION THREAD

**Discussion by RJ-Eckie** [SynthstromAudible/DelugeFirmware#2905] [General]
This is basically for discussing changes to existing code, before making them. This goes together with #2904. I've been doing some cleanup on `SessionView` (now `SongView`) but to make any real difference I quickly realized cleanup needs to start at the very bottom, I think in `deluge.cpp`. (scary) 

- So first a very basic question: do we feel that an object & inheritance-oriented kind of structure is something to aim for? (I sure do.) If so, I would start with wrapping everything from `deluge.h` and `deluge.cpp` in a `class Deluge`, and go from there. (Except the C stuff, I'm personally not brave enough to touch that.)
- There are a bunch of `#ifdef` codeblocks in `deluge.cpp` that almost exclusively deal with tests. I read somewhere that a certain test-mode was no longer worked on, but I'm not sure if this is what that was talking about. Does anyone know the ins and outs of these testblocks? It would create a lot of space if we could get rid of them. (And if not, possibly move them to a separate class/file.)
- Same question for `SPAM_MODE`, do we know what this is?
- There is one point in `deluge.cpp` where a choice is made between either running in a linear loop or in a task manager. Is this still an active option, or has a definitive choice been made about how the software is run?

Thanks

**Comment by stellar-aria** [comment]
The contents of deluge.cpp cannot be wrapped in a class/namespace or be member functions as they have to have C linkage for calling outside of the Deluge application code, for example from `main.c`.

`deluge/deluge.cpp` can be thought of as the traditional application-level `main.cpp` file, with associated `main` function (in this case, `deluge_main`). `main.c` instead, is our OS-level entrypoint, to be called from our first-stage-init entrypoint (`resetprg`) and in turn from assembly (`reset_handler.S`)

The boot process is therefore like this:
`start.S->reset_handler.S->resetprg()->main()->deluge_main()`

The tester macros have not been used since community release.
Ditto for the spam stuff.

We currently always build using the task manager, but considering it's a relatively new thing, it might be better to keep those compiler defines so that we can do fast A/B comparisons in certain situations. @m-m-adams would be the one to talk to about that.

**Reply by RJ-Eckie** [reply]
Thanks for the feedback!

Yeah so basically what I think would be helpful is to leave all the `extern "C"` declarations in place so that they can be called unchanged from the C code, but to wrap as many variables and methods as possible (that now live globally) inside a `Deluge` class. This would leave all low-level interactions in place, while still providing an ultimate base-class that we can have the c++ code build upon.

_(**Edit:** even without initially wrapping anything, just creating a `Deluge` class could give us a starting point to start moving some code into a clearly structured and documented hierarchy.)_

For now, tester and spam stuff could probably at least be organized into a separate file which can be in/excluded with a single define. This would increase legibility a bunch.

`registerTasks()` and `mainLoop()` are quite comprehensible as is so they could just be left untouched.

**Reply by stellar-aria** [reply]
yeah, that sounds reasonable, I'd say go for it.

**Reply by stellar-aria** [reply]
One of the huge problems with the codebase as-is, is the very prolific and widely distributed global and/or static variables. A half a year or so we also threw around the idea of a big ball of goo state object that just collected all of those variables, and considering this new class will be trying to capture application world-state it might be relevant to revisit the idea.

**Reply by RJ-Eckie** [reply]
I think my strategy would be a little bit different. I agree the goal of `Deluge` would be to ultimately capture the full state of everything, but I'm not sure if big-ball-of-goo'ing all global variables would help us a lot towards making the codebase more efficient and easier to navigate?

Instead of trying to capture everything at once, I would bit-by-bit start moving code from being global to being owned by `Deluge`. All public code is accessible from `Deluge` by default. Wherever public code needs to access something that has been moved from being public to being owned by `Deluge`, it can be structured in a way that it is statically available. _(F.e.: large systems like the file system should still be statically available and can reveal previously-public variables and methods.)_

This would result in a state where the existing code with global methods and variables lives alongside a growing, organized `Deluge` class. This gives us the chance to know what is organized and what isn't, to work on step-by-step restructuring without having to make sudden large fundamental changes to the existing code, and to organise the most essential systems first. Realistically, some public code might never be captured in `Deluge`, and one could argue that isn't even that big of an issue.

**Reply by m-m-adams** [reply]
I'm not sure that deluge.cpp is really the best place to start, there's very little in that file that makes sense as an object. It's just the direct hardware interfaces which are fundamentally global (e.g. there's only one spi bus) and the functions with C linkage that need to be kept as is

**Comment by m-m-adams** [comment]
For the other questions:

Spam mode can probably be removed, I think it's for hardware tests with the PIC?

The ifdef tests can be removed, new tests are in their own directory with mocking to run on your own computer

Moving to more object oriented code is good where it makes sense, but I personally prefer C style code where it's interfacing with hardware or C code. Also be aware that virtual dispatch takes an appreciable amount of time and should be avoided in the hot path (i.e. audio rendering). We're generally aiming for generic code with template parameters there instead - see the modfx or filters for examples
